
# perform LAVA analysis for all 100 replicates for a given block and condition, and for a specific reference data set
# - for faster analysis, each replicate for each phenotype is loaded in at once and processed through a single process.locus() call
# - bivariate analysis is then performed for each matched pair of phenotypes for each replicate
#   - param.lim is set to Inf to turn off discarding estimates that would be too far out of bounds, in order to obtain better estimates of statistical efficiency
#   - adap.thresh is set to NULL to turn off the adaptive p-value computation, since the default 10,000 simulations gives sufficient resolution given the significance thresholds used later

# arguments:
# - [1] reference data set
# - [2] condition label
# - [3] block ID

# input files:
# - reference genotype data PLINK format files: stored per block, subset to a specific number of individuals (for UK Biobank) number of individuals, containing only the SNPs to be used for that block
#   - ukb_main: same data as used in generate.r, N = 167636
#   - ukb_503 / ukb_1K / ukb_5K / ukb_10K / ukb_100K: UK Biobank subsamples of different sizes; do not overlap with ukb_main
#   - g1000: European panel of the 1,000 Genomes data
# - LAVA-specific reference LD files for chromosome 22 (lava_ld)
# - .info and .stats1/2 files created by generate.r script

# output files:
# - .univ and .bivar files with analysis results



flags = commandArgs(T)

library(LAVA)

reference = flags[1]
condition = flags[2]
block = as.numeric(flags[3])


root = "[ROOT DIRECTORY]"
input.dir = paste0(root, "/generate/output/", condition, "/block", block)  # contains the files generated by the generate.r script, for the given block and condition

if (reference == "lava_ld") {
	ld.root = "[LD ROOT DIRECTORY]"
	pkg.dir = paste0(ld.root, "/LAVA-0.1.5")
	pkgload::load_all(pkg.dir)  # load LAVA v0.1.5
	
	ref.dir = paste0(ld.root, "/lava-ukb")  # contains the reference data LD files
	ref.prefix = paste0(ref.dir, "/lava-ukb-v1.1_chr22")
} else {
	library(LAVA)  # load LAVA v0.1.0
	ref.dir = paste0(root, "/blocks/", reference)  # contains the reference data genotype files
	ref.prefix = paste0(ref.dir, "/", reference, "_block", block) 
}


working.dir = paste0("temp-", reference, "_", condition, "_block", block)
output.dir = paste0(root, "/analyse/lava/output/", reference, "/", condition)

dir.create(working.dir, showWarnings=F, recursive=T)
dir.create(output.dir, showWarnings=F, recursive=T)


# process the bulk summary statistics into separate input files, and create a corresponding phenotype info file
snp.info.file = paste0(input.dir, "/", condition, "_block", block, ".info")
snp.stats.files = paste0(input.dir, "/", condition, "_block", block, ".stats", 1:2)
info.file = paste0(working.dir, "/phenotypes.info")

snp.info = read.table(snp.info.file, header=T, stringsAsFactors=F)
snp.stats = lapply(snp.stats.files, read.table, header=T, stringsAsFactors=F)
no.draws = ncol(snp.stats[[1]])

info = data.frame(phenotype=paste0("pheno", rep(1:2, each=no.draws), "_draw", 1:no.draws), cases=NA, controls=NA)
info$filename = paste0(working.dir, "/", info$phenotype, ".stats")
write.table(info, file=info.file, row.names=F, quote=F, sep="\t")

for (p in 1:2) {
	for (i in 1:no.draws) {
		curr = cbind(snp.info, Z=snp.stats[[p]][,i])
		write.table(curr, file=paste0(working.dir, "/pheno", p, "_draw", i, ".stats"), row.names=F, quote=F, sep="\t")
	}
}


# process input files for analysis, and create a block info object
input = process.input(input.info.file=info.file, sample.overlap.file=NULL, ref.prefix=ref.prefix)
locus.info = list(LOC=paste0("block", block), SNPS=snp.info$SNP)

# process locus for all 200 phenotypes simultaneously, and perform univariate testss
locus = process.locus(locus.info, input, drop.failed=F)
univ = run.univ(locus)


# perform bivariate analysis
bivar = NULL; failed = rep(F, no.draws)
for (i in 1:no.draws) {
	failed[i] = any(locus$failed[c(0,100) + i])
	if (!failed[i]) {
		b = suppressWarnings(try(run.bivar(locus, phenos=paste0("pheno", 1:2, "_draw", i), adap.thresh=NULL, param.lim=Inf), silent=T)) 
		if (class(b) != "try-error") bivar = rbind(bivar, cbind(index=i, failed=F, b))
		else failed[i] = T
	}
}

# add in dummy NA results lines for replicates where bivariate analysis was not possible to simplify processing later
if (any(failed)) {
	no.failed = sum(failed)
	add = bivar[rep(1, no.failed), ,]; add[,] = NA; rownames(add) = NULL
	add$index = which(failed)
	add$failed = T
	add$phen1 = paste0("pheno1_draw", which(failed))
	add$phen2 = paste0("pheno2_draw", which(failed))
	
	bivar = rbind(add, bivar)
	bivar = bivar[order(bivar$index),]
}


# save output
write.table(univ, file=paste0(output.dir, "/", condition, "_block", block, ".univ"), row.names=F, quote=F, sep="\t")
write.table(bivar, file=paste0(output.dir, "/", condition, "_block", block, ".bivar"), row.names=F, quote=F, sep="\t")

unlink(working.dir, recursive=T)










